### DISCUSSION TITLE
NPM Package for meta script?

### DISCUSSION THREAD

**Discussion by Alex-Wenner-FHR** [h2oai/wave#1086] [Q&A]
Is it possible to npm install a package to use for meta script? I see in the meta callback inline script example, a cdnjs.cloudflare package is used. Just wondering if it would be possible to use an npm package and if so, how would one do that?

**Comment by mturoci** [comment]
>I see in the meta callback inline script example, a cdnjs.cloudflare package is used

You can basically use any script file that is hosted anywhere on the web - just like native HTML.

>Just wondering if it would be possible to use an npm package and if so, how would one do that?

In general, the best way to go with this is to look if the package you need is not also made available from a CDN. Most serious packages provide it.

If you however need a package that is only available via npm, you will need some hacking, which is not recommended:

1. npm install your package
2. Open node_modules and see how it is packaged. For example React comes with 2 types of packaging - umd and commonJs modules
![image](https://user-images.githubusercontent.com/64769322/138648926-6904caf2-c68b-4870-a48c-083d9657b3bf.png)
so you pick commonjs minified production bundles.
3. You upload them to your wave content server (`q.site.upload`) - this way you make "your own CDN"
4. Serve them via meta script
5. Profit

Note that this is hacky and should almost never be used this way. Wondering what npm package you want to use?

**Reply by Alex-Wenner-FHR** [reply]
Ah yeah that does not feel like a good method to practice. Thanks for the example on how to do this, but I think I am going to stay away from that! I am browsing for a pivot table package and found https://www.npmjs.com/package/quick-pivot … will be delighted when the wave solution makes its way into the scene!
